dirty_cat/string_distances.py

```diff
@@ -19,6 +20,52 @@
 # TODO vectorize these functions (accept arrays)
 
 
+def get_ngram_count(X, ngram_range):
+    """Compute the number of ngrams in a string.
+
+    Here is where the formula comes from:
+
+    * the number of 3-grams in a string is the number of sliding windows of
+      size 3 in the string: len(string) - 3 + 1
+    * this can be generalized to n-grams by changing 3 by n.
+    * when given a ngram_range, we can sum this formula over all possible
+      ngrams
+
+    """
+    min_n, max_n = ngram_range
+    ngram_count = 0
+
+    for i in range(min_n, max_n + 1):
+        ngram_count += len(X) - i + 1
+
+    return ngram_count
+
+
+def preprocess(x):
+    """Combine preprocessing done by CountVectorizer and the SimilarityEncoder.
+
+    Different methods exist to compute the number of ngrams in a string:
+
+    - Simply sum the values of a count vector, which is the ouput of a
+      CountVectorizer with analyzer="char", and a specific ngram_range
+    - Compute the number of ngrams using a formula (see ``get_ngram_count``)
+
+    However, in the first case, some preprocessing is done by the
+    CountVectorizer that may change the length of the string (in particular,
+    stripping sequences of 2 or more whitespaces into 1). In order for the two
+    methods to output similar results, this pre-processing is done upstream,
+    prior to the CountVectorizer.
+    """
+
+    # preprocessing step done in ngram_similarity
+    x = ' %s ' % x
+
+    # preprocessing step done in the CountVectorizer
+    _white_spaces = re.compile(r"\s\s+")
+
+    return _white_spaces.sub(' ', x)
```

Comment:
Can we turn off this preprocessing in count vectorizer? We don't really want it, and it probably slows things down.

Reply:
nope!

Reply:
:(